c/foreign.c

```diff
@@ -228,9 +229,63 @@ static void load_shared_object(path) const char *path; {
 
     tc_mutex_acquire()
 
-    handle = dlopen(path, RTLD_NOW);
-    if (handle == (void *)NULL)
-        S_error2("", "(while loading ~a) ~a", Sstring_utf8(path, -1), s_dlerror());
+#if defined(HPUX)
+    /*
+     * With NULL path, use RTLD_SELF to act like dlopen(NULL, ...)
+     * See: https://docstore.mik.ua/manuals/hp-ux/en/B2355-60130/dlsym.3C.html
+     */
+    if (!path) {
+        handle = (void *)RTLD_SELF;
+    }
+    else
+#elif defined(WIN32)
+    /*
+     * With NULL path, manually add loaded modules to the list of shared objects.
+     */
+    if (!path) {
+        size_t i, N;
+        DWORD ret;
+        HMODULE modules_stack[1024];
+        DWORD cur_num_bytes = sizeof modules_stack;
+        DWORD req_num_bytes;
+        HMODULE *modules = modules_stack;
+        HMODULE *dynamic_storage = NULL;
+
+        /* Always try to use the stack, allocate memory only if necessary */
+        ret = EnumProcessModules(GetCurrentProcess(), modules, cur_num_bytes, &req_num_bytes);
+        while (ret && req_num_bytes > cur_num_bytes) {
```

Comment:
A loop is used there, rather than a simple `if`, to be resilient in the face of weird foreign threads that loads DLL concurrently and independently of the Scheme system. This loop _ought_ to terminate quite fast, otherwise that means something bad happens anyway because it would mean the list of loaded modules grows very fast, and that's in itself the biggest concern, _not that allocation loop_.